Sources/System/FileHelpers.swift

```diff
@@ -123,3 +123,34 @@ extension FileDescriptor {
     }
   }
 }
+
+extension FileDescriptor.Pipe {
+  /// Runs a closure and then closes both ends of this pipe, even if an error occurs.
+  ///
+  /// - Parameter body: The closure to run.
+  ///   If the closure throws an error,
+  ///   this method closes both ends of this pipe before it rethrows that error.
+  ///
+  /// - Returns: The value returned by the closure.
+  ///
+  /// If `body` throws an error
+  /// or an error occurs while closing the file descriptor,
+  /// this method rethrows that error.
+  public func closeAfter<R>(_ body: () throws -> R) throws -> R {
```

Comment:
There are two scenarios where this is useful:
1. You are forking off **two** child processes and simply connecting them with a pipe
2. In an async/await world, you will be able to write or read some data entirely within the `closeAfter` block

Reply:
For scenario 1, are you handing off one FD to a child and another to the other child, or do both get the full pipe?

For scenario 2, could you give more context?

If we are making a proper pipe type, does it make sense to have read/write operations on it that would hook up to the appropriate end?

Reply:
> For scenario 1, are you handing off one FD to a child and another to the other child, or do both get the full pipe?

One to each, the simplest example of this is implementing something like `$ echo foo | sed 's/foo/bar/'` in Swift. The naive way to solve this without `closeAfter` is wrong:
```
let (input, output) = pipe
try input.closeAfter {
  try launchChildProcess(input: input)
}
try output.closeAfter {
  try launchChildProcess(output: output)
}
```
The issue here is that if the first `launchChildProcess` fails, `output` will not be closed. If we do `closeAfter` on the pipe, this behaves correctly.

> For scenario 2, could you give more context?

With #54, I'm imagining something like:
```
let output: String = try await Pipe().closeAfter { pipe in
  try launchChildProcess(output: pipe.input)
  return try await pipe.output.readToEOF()
}
```

> If we are making a proper pipe type, does it make sense to have read/write operations on it that would hook up to the appropriate end?

Perhaps? I don't see a huge benefit to `Pipe.read` over `Pipe.output.read`, though perhaps it will discourage less sensible usage like `Pipe.output.write`. Either way, this can be added as a subsequent improvement.

**ALSO** My first code example is actually wrong, and launching child processes should look something like this `launchChildProcess(input: output)`. Once we actually get around to implementing something along the lines of `posix_spawn` we can do the same thing `Process` does and accept either a `FileDescriptor` or a `Pipe` and in the case that a `Pipe` is provided, the process-launching code will select the correct end of the pipe to use depending on whether it is passed for stdin/stdout or stderr.

Reply:
> Once we actually get around to implementing something along the lines of `posix_spawn` we can do the same thing `Process` does and accept either a `FileDescriptor` or a `Pipe` and in the case that a `Pipe` is provided, the process-launching code will select the correct end of the pipe to use depending on whether it is passed for stdin/stdout or stderr. 

This is a good argument.

`Process` adds a large number of high-level assumptions that aren't built into `posix_spawn` -- for instance, `posix_spawn` does not assume that fd 0 must be an input file descriptor. (It most often is, but it's just a convention -- in its direct syscall wrappers, System probably should not introduce assumptions that would prevent its use in niche cases.) Without that assumption, a helper that takes a pipe for fd 0 is less obviously desirable.

But this is a nitpicky argument -- whatever wrapper we create for `posix_spawn_file_actions_t`, it seems desirable to have a set of convenience methods to set up conventional stdin/stdout/stderr channels.